logic/LogicManager.java

```diff
@@ -106,12 +107,21 @@ private FunctionalIterator<Rule> rulesWithNegations() {
         return rules().filter(rule -> !rule.when().negations().isEmpty());
     }
 
-    public FunctionalIterator<Unifier> unifiers(Concludable concludable, Rule rule) {
-        return iterate(applicableRules(concludable).getOrDefault(rule, new HashSet<>()));
+    public Map<Rule, Set<Unifier>> applicableRules(Concludable concludable) {
+        Map<Rule, Set<Unifier>> unifiers = logicCache.unifiers().getIfPresent(concludable);
+        if (unifiers == null) {
+            unifiers = concludable.computeApplicableRules(conceptMgr, this);
+        }
+        return unifiers;
     }
 
-    public Map<Rule, Set<Unifier>> applicableRules(Concludable concludable) {
-        return logicCache.applicableRules().get(concludable, c -> c.applicableRules(conceptMgr, this));
+    public void indexConcludables(ResolvableConjunction conjunction) {
+        link(Iterators.single(conjunction),
+                iterate(conjunction.negations()).flatMap(negation -> iterate(negation.disjunction().conjunctions())))
+                .flatMap(conj -> iterate(conj.concludables())).
+                forEachRemaining(concludable -> {
+                    logicCache.unifiers().get(concludable, c -> c.computeApplicableRules(conceptMgr, this));
+                });
     }
```

Comment:
Making the indexing explicit.
We do pay the cost of iterating through the conjunction even if it is cached.